src/algebraic_geometry/Scheme.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import algebraic_geometry.locally_ringed_space
+import algebraic_geometry.Spec
+
+/-!
+# The category of schemes
+
+A scheme is a locally ringed space such that every point is contained in some open set
+where there is an isomorphism of presheaves between the restriction to that open set,
+and the structure sheaf of `Spec R`, for some commutative ring `R`.
+
+A morphism is schemes is just a morphism of the underlying locally ringed spaces.
+
+-/
+
+open topological_space
+open category_theory
+open Top
+
+namespace algebraic_geometry
+
+/--
+We define `Scheme` as a `X : LocallyRingedSpace`,
+along with a proof that every point has an open neighbourhood `U`
+so that that the restriction of `X` to `U` is isomorphic, as a space with a sheaf of commutative rings,
+to `Spec.SheafedSpace R` for some `R : CommRing`.
+
+(Note we're not asking in the definition that this is an isomorphism as locally ringed spaces,
+although that is a consequence.)
+-/
+structure Scheme extends X : LocallyRingedSpace :=
+(local_affine : ∀ x : carrier, ∃ (U : opens carrier) (m : x ∈ U) (R : CommRing)
+  (i : X.to_SheafedSpace.restrict CommRing _ (opens.inclusion_open_embedding U) ≅ Spec.SheafedSpace R), true)
+
+-- PROJECT
+-- In fact, we can construct `Spec.LocallyRingedSpace R`,
+-- and the isomorphism `i` above is an isomorphism in `LocallyRingedSpace`.
+-- However this is a consequence of the above definition, and not necessary for defining schemes.
+-- We haven't done this yet because:
+-- 1. We haven't proved that the stalk of the structure sheaf is isomorphic to the localisation
+--    **as a ring**, only at the level of `Type`.
+--    To do this, we need to know that `forget CommRing` preserves filtered colimits.
+-- 2. We haven't shown that you can restrict a `LocallyRingedSpace` along an open embedding.
+--    We can do this already for `SheafedSpace` (as above), but we need to know that
+--    the stalks of the restriction are still local rings, which we follow if we knew that
+--    the stalks didn't change.
+--    This will follow if we define cofinal functors, and show precomposing with a cofinal functor
+--    doesn't change colimits, because open neighbourhoods of `x` within `U` are cofinal in
+--    all open neighbourhoods of `x`.
+
+namespace Scheme
+
+/--
+Every `Scheme` is a `LocallyRingedSpace`.
+-/
+-- (This parent projection is apparently not automatically generated because
+-- we used the `extends X : LocallyRingedSpace` syntax.)
+def to_LocallyRingedSpace (S : Scheme) : LocallyRingedSpace := { ..S }
+
+/--
+The empty scheme, as `Spec 0`.
+-/
+noncomputable
+def empty : Scheme :=
+{ local_ring := λ x, false.elim (prime_spectrum.punit x),
+  local_affine := λ x, false.elim (prime_spectrum.punit x),
+  ..Spec.SheafedSpace (CommRing.of punit) }
```

Comment:
How else would I do it?

Reply:
I thought about this a bit more and my method means more work in other places. You can *explicitly* define a sheaf of rings on the empty set: send empty to 0 and that's it. Now unfortunately you have to check that this is a sheaf, which involves checking that the map from 0 to an arbitrary product of 0's is an isomorphism (this is what we did in the perfectoid project, where we had to prove also that it was a homeomorphism of topological rings :-) ). But after that it's easier -- it's local because there are no points, and every point has an affine neighbourhood because there are no points. My observation is simply that this approach does not ever need the Spec construction to ever be applied. However the advantage of your approach is that Spec(0) is automatically a sheafed space because of earlier work. Ultimately your approach is shorter, given what you have, so it's probably saner.

Reply:
In fact, I think what I've done here is just as easy!

```
local_ring := λ x, false.elim (prime_spectrum.punit x),
```
is precisely saying "it's local because there are no points", and
```
local_affine := λ x, false.elim (prime_spectrum.punit x),
```
is saying "every point has an affine nbhd because there are no points".

I guess what you mean is that we had to prove `prime_spectrum.punit` in the first place.